### PR TITLE
[FLINK-18750][table] SqlValidatorException thrown when select from a …

### PR DESCRIPTION
**Pull request by danny0405**
…view which contains a UDTF call

## What is the purpose of the change

This patch fix the problem that a table function can not reference the `JOIN` left items for `CREATE VIEW` query.

## Brief change log

  - Fix the issue in blink planner
  - Add UT

## Verifying this change

Added UT.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

**Comment by danny0405** [comment]
@dawidwys Can you take a look for this fix, thanks in advance ~

**Comment by dawidwys** [comment]
Are we planning to bump Calcite version for 1.12? @wuchong @twalthr 

If we do, I'd prefer we do not add this workaround. If we don't the PR looks good to me

**Comment by danny0405** [comment]
> Are we planning to bump Calcite version for 1.12? @wuchong @twalthr
> 
> If we do, I'd prefer we do not add this workaround. If we don't the PR looks good to me

I'm afraid there is no time for the Calcite version bump for 1.12, the recent versions of Calcite made many breaking changes and it is very probably introduces bugs, need a full local test before we decide if to upgrade.

**Comment by twalthr** [comment]
We will release Flink 1.12 end of October. We should bump the Calcite version until then.

**Comment by twalthr** [comment]
Could you name a few of these breaking changes? I know about the dropped windows but what else?

**Comment by danny0405** [comment]
> Could you name a few of these breaking changes? I know about the dropped windows but what else?

I listed some features i though important (very probably breaking) here:

- [CALCITE-3786] The rel node digest was refactored
- [CALCITE-3896 Top down trait pass through
- [CALCITE-3916 Support cascades style top-down driven rule apply
- [CALCITE-2569] UDFs that are table functions must implement SqlTableFunction and have CURSOR as their return type
- [CALCITE-3923] Refactor how planner rules are parameterized

But anyway, Calcite is a single open source community and we can not always depend on the version upgrade, and Calcite does not always evolve following Flink's request. If we have a simple workaround for a quick fix(log an issue there and remove the code when upgrade), it is better than just wait for the upgrade.

This fix is critical, use can not even use the table function now in a `CREATE VIEW` and is a blocker for VVP platform.

**Comment by dawidwys** [comment]
> If we have a simple workaround for a quick fix(log an issue there and remove the code when upgrade), it is better than just wait for the upgrade.

I mostly agree. I am a bit sceptical though in that particular case for the following reason:
1. That's not a simple fix, that will be backported back to Calcite. It is a workaround that disables a feature (expanding identifiers) for the `LATERAL` to work.
2. The fix is not in Calcite classes. We must remember to remove this workaround when upgrading Calcite. It's not as simple as dropping Calcite classes.
3. We do not need to wait for Calcite to implement that. If I understand it correctly, the fix is already released.

**Comment by dawidwys** [comment]
To emphasize it again though, personally if we don't want to upgrade calcite, personally I am fine with that change.

**Comment by wuchong** [comment]
I'm fine with this hot fix. However, personally, I think we should catch up Calcite versions. Breaking changes and the upgrading effort is not a strong reason to stop us to upgrade. Otherwise,  the cost of upgrading will become greater and greater.

**Comment by twalthr** [comment]
Merging hot fixes is fine with me as well. But we should have a strategy when and how such a hotfix will be removed from the Flink code base again.

**Comment by danny0405** [comment]
Finally i have added a new component named `Expander` for the sql identifier expanding, so that we do not need to depend on the Calcite upgrade, also this would avoid the expanding logic to be buggy again and again.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/utils/Expander.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+	public static Expander create(FlinkPlannerImpl planner) {
+		return new Expander(planner);
+	}
+
+	/** Expands identifiers in a given SQL string, returning a {@link Expanded}. */
+	public Expanded expanded(String ori) {
+		final Map<SqlParserPos, SqlIdentifier> identifiers = new HashMap<>();
+		final SqlNode oriNode;
+		final SqlNode validated;
+		oriNode = planner.parser().parse(ori);
```
Comment (dawidwys) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/utils/Expander.java:75]: nit: inline assignment

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/utils/Expander.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+		// parse again because validation is stateful, that means the node tree was probably
+		// mutated.
+		validated = planner.validate(planner.parser().parse(ori));
```
Comment (dawidwys) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/utils/Expander.java:78]: nit: inline assignment

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/utils/Expander.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+		 * given unparse function. */
+		public String substitute(Function<SqlNode, String> fn) {
+			final SqlShuttle shuttle = new SqlShuttle() {
+				@Override
+				public SqlNode visit(SqlIdentifier id) {
+					if (id.isStar()) {
+						return id;
+					}
+					final SqlIdentifier toReplace = identifiersMap.get(id.getParserPosition());
+					if (toReplace != null && id.names.size() >= toReplace.names.size()) {
```
Comment (dawidwys) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/utils/Expander.java:116]: Shouldn't this be rather:
```
if (toReplace == null || id.names.size() >= toReplace.names.size()) {
```
?

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/utils/Expander.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+ *
+ * <p>Calling {@link Expanded#toString()} generates a string that is similar to
+ * SQL where a user has manually converted all identifiers as expanded, and
+ * which could then be persisted as expanded query of a Catalog view.
+ *
+ * <p>For more advanced formatting, use {@link Expanded#substitute(Function)}.
+ *
+ * <p>Adjust {@link SqlParser.Config} to use a different parser or parsing options.
+ */
+public class Expander {
```
Comment (dawidwys) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/utils/Expander.java:58]: Will it expand function identifiers as well? Can we have a test for that case?

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/utils/Expander.java (hunk 3 above)
Comment (danny0405) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/utils/Expander.java:116]: Yes, nice catch ~

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/utils/Expander.java (hunk 4 above)
Comment (danny0405) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/utils/Expander.java:58]: I'm afraid we never support expanding functions, i have added a test.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/optimize/program/FlinkChainedProgram.scala [hunk 5]
```diff
@@ -62,8 +62,8 @@ class FlinkChainedProgram[OC <: FlinkOptimizeContext]
         val result = program.optimize(input, context)
         val end = System.currentTimeMillis()
 
-        if (LOG.isDebugEnabled) {
-          LOG.debug(s"optimize $name cost ${end - start} ms.\n" +
+        if (true) {
```
Comment (dawidwys) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/optimize/program/FlinkChainedProgram.scala:65]: unrelated change

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/utils/Expander.java (hunk 4 above)
Comment (dawidwys) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/utils/Expander.java:58]: The test you added uses a built-in function, which has a single part identifier which does not differ from not expanded one. Are you saying we never expanded identifiers for UDFs?

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/optimize/program/FlinkChainedProgram.scala (hunk 5 above)
Comment (danny0405) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/optimize/program/FlinkChainedProgram.scala]: Oops, i committed some test code, fixed.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/utils/Expander.java (hunk 4 above)
Comment (danny0405) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/utils/Expander.java:58]: The first test also add a UDTF there and it still no expanded.

I didn't find any evidence that Calcite validator expands the function IDs, the `DelegatingScope#fullyQualify` expands the table columns, the `IdentifierNamespace#validateImpl` expands the table name.

Or did i miss something ?

Comment (dawidwys) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/utils/Expander.java:58]: I don't know where it happens but a test like:
```
  @Test
  def testViewExpandingWithUDF(): Unit = {
    val tableUtil = tableTestUtil(this)
    val tableEnv = tableUtil.tableEnv
    tableEnv.createTemporaryFunction("func", classOf[PrimitiveScalarFunction])
    val createView =
      """
        |CREATE VIEW tmp_view AS
        |  SELECT func(1)
        |""".stripMargin
    tableEnv.executeSql(createView)
    val objectID = ObjectIdentifier.of(tableEnv.getCurrentCatalog,
      tableEnv.getCurrentDatabase, "tmp_view")
    val view: CatalogBaseTable = tableEnv.getCatalog(objectID.getCatalogName)
      .get().getTable(objectID.toObjectPath)
    assertThat(view.asInstanceOf[CatalogView].getExpandedQuery, is("SELECT `default_catalog`.`default_database`.`func`(1)"))
  }
```

passes on master and fails with your changes.

Comment (danny0405) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/utils/Expander.java:58]: Yes, the `BridgingSqlFunction` did the function id expanding out of the scope of the `SqlValidator`, so in the code base of this patch, it does not work because `BridgingSqlFunction` id position was always `SqlParserPos.ZERO` so it can not match the original identifier.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/utils/Expander.java [hunk 6]
```diff
@@ -70,30 +75,43 @@ public static Expander create(FlinkPlannerImpl planner) {
 	/** Expands identifiers in a given SQL string, returning a {@link Expanded}. */
 	public Expanded expanded(String ori) {
 		final Map<SqlParserPos, SqlIdentifier> identifiers = new HashMap<>();
+		final Map<String, SqlIdentifier> funcNameToId = new HashMap<>();
```
Comment (dawidwys) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/utils/Expander.java:78]: Why did you decide to use a different key for functions? Can't we use the `SqlParserPos` as well?

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/common/ViewsExpandingTest.scala [hunk 7]
```diff
@@ -20,9 +20,12 @@ package org.apache.flink.table.planner.plan.common
 
 import org.apache.flink.api.scala._
 import org.apache.flink.table.api._
-import org.apache.flink.table.catalog.{CatalogView, CatalogViewImpl, ObjectPath}
-import org.apache.flink.table.planner.utils.{TableTestBase, TableTestUtil, TableTestUtilBase}
+import org.apache.flink.table.catalog.{CatalogBaseTable, CatalogView, CatalogViewImpl, ObjectIdentifier, ObjectPath}
+import org.apache.flink.table.planner.runtime.stream.sql.FunctionITCase.PrimitiveScalarFunction
```
Comment (dawidwys) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/common/ViewsExpandingTest.scala:24]: Let's not import the `PrimitiveScalarFunction`. I don't think it makes sense to introduce a cross dependency between the tests. I think it is easy enough to add a simple function in this class.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/utils/Expander.java (hunk 6 above)
Comment (danny0405) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/utils/Expander.java:78]: Because the `BridgingSqlFunction` lost the parser position information.

**Comment by dawidwys** [comment]
Thanks for the update. LGTM
